ansible_base/authentication/utils/authentication.py

```diff
@@ -0,0 +1,116 @@
+        logger.info(f"Authenticator {authenticator.name} already authenticated {uid} as {new_username}")
+        return new_username
+
+    # We didn't get an exact match. If any other provider is using this uid our id will be uid<hash>
+    if AuthenticatorUser.objects.filter(uid=uid).count() != 0:
+        # Some other provider is providing this username so we need to create our own username
+        new_username = get_local_username({'username': uid})
+        logger.info(
+            f'Authenticator {authenticator.name} wants to authenticate {uid} but that'
+            f'username is already in use by another authenticator,'
+            f' the user from this authenticator will be {new_username}'
+        )
+        return new_username
+
+    # We didn't have an exact match but no other provider is servicing this uid so lets return that for usage
+    logger.info(f"Authenticator {authenticator.name} is is able to authenticate user {uid} as {uid}")
+    return uid
+
+
+def get_or_create_authenticator_user(
+    username: str, authenticator: Authenticator, user_details: dict = dict(), extra_data: dict = dict()
+) -> Tuple[Optional[AbstractUser], Optional[AuthenticatorUser], Optional[bool]]:
+    """
+    Create the user object in the database along with it's associated AuthenticatorUser class.
+    In some cases, the user may already be created in the database.
+    This assumes that the given user_name has already been resolved using the determine_username_from_uid but will do a quick double check
+
+    This should be called any non-social auth plugins.
+
+    Inputs
+    username: The username we are going to be created
+    user_details: Any details about the user from the source (first name, last name, email, etc)
+    authenticator: The authenticator authenticating the user
+    extra_data: Any additional information about the user provided by the source.
+                For example, LDAP might return sn, location, phone_number, etc
+    """
+
+    created = None
+    try:
+        # First see if we have an auth user and if so update it
+        auth_user = AuthenticatorUser.objects.get(uid=username, provider=authenticator)
+        auth_user.extra_data = extra_data
+        auth_user.save()
+        created = False
+    except AuthenticatorUser.DoesNotExist:
+        # Ensure that this username is not already tired to another authenticator
+        auth_user = AuthenticatorUser.objects.filter(uid=username).first()
+        if auth_user is not None:
+            logger.error(
+                f'Authenticator {authenticator.name} attempted to create an AuthenticatorUser for {username}'
+                f' but that id is already tied to authenticator {auth_user.provider.name}'
+            )
+            return None, None, None
+
+    # ensure the authenticator isn't trying to pass along a cheeky is_superuser in user_details
+    details = {k: user_details.get(k, "") for k in ["first_name", "last_name", "email"] if k}
```

Comment:
1. The `if k` condition is still here, seems like a mistake. It's either meant to do something else or it simply can be deleted.
2. Minor(nit): Personally I'd move this list to a constant:

```python
USER_DETAILS_FIELDS = ["first_name", "last_name", "email"] 

# ...

def get_or_create_authenticator_user(...):
    # ...
    details = {k: user_details.get(k, "") for k in USER_DETAILS_FIELDS}
```

Reply:
I thought about moving it to a constant but we don't use it anywhere else so I was just going to put it inline. 
The goal of this was to make sure we only accept specific fields out of the user_fields instead of users being able to put junk in. I don't remember why the `if` was there but yea, it seems useless at this point in this context.